Upgrade observer/replay interface features. #5121

rob-v opened the pull request:
Statistics window restyled, each tab has hotkey configurable in Settings dialog. Scrollbar added when there is more than 8 players. F1 - show/hide Statistics (previous shown tab is displayed).
In TD there is added number of currently produced items in Production tab. After these changes, more UI improvements can be done simpler.

![obrazok](https://cloud.githubusercontent.com/assets/16348750/24326202/9a22baa8-11a9-11e7-9183-9ab5d3ff7a01.png)
![obrazok](https://cloud.githubusercontent.com/assets/16348750/24326204/9e1a79c0-11a9-11e7-924c-2d87e20137aa.png)

![obrazok](https://cloud.githubusercontent.com/assets/16348750/24326223/206d8232-11aa-11e7-8f3d-521aa00b564b.png)

Commit:
Upgrade observer/replay interface features. #5121

GraionDilach commented on OpenRA.Mods.Common/Widgets/ObserverProductionIconsWidget.cs:
```diff
@@ -77,43 +77,88 @@ public override void Draw()
+			}
+		}
+
+		int SortGroupByInfoType(ProductionItem queue)
```
WHAT?!

rob-v replied:
Something more useful?

GraionDilach replied:
Arbitrary hardcode. I have no idea where this was sued or why this was even added but this flatout goes against the direction OpenRA's supposed to take lately.

GraionDilach replied:
IOW: this probably masks a bug instead of resolving an issue.

rob-v replied:
That SortGroupByInfoType can be removed, it is optional. I don't like it also. There was comment to display items in this order - Building, .. Infrantry, Vehicle, Aircraft. Do you have an idea how to sort it in general?

GraionDilach replied:
I don't see what was the issue with the legacy sort-queues-in-their-order-of-enabling. If that has issues with mods, that's the mod's fault.

rob-v replied:
For D2K someone mentioned it was in different other than other mods. I could keep it like it was, however it will be now more difficult with that more general code for multique support. As there are many buildings in various orders and I implemented to display in this order:
building, defence, infantry, vehicle. With sort-queues-in-their-order-of-enabling it won't be sorted.

GraionDilach reviewed (changes requested):
This PR smells of design issues, ignores the issue of third-party mods ending up with overlapped support powers and production issues because the end result "looks nicer" in the shipped mods.

rob-v commented:
Only SortGroupByInfoType smells and can be removed. Otherwise it is 'general' like it was so far. I expect it can be further improved. It is written in general way - any mod with single or multique will work - like currently it works for RA and TD. So it is better than it was.

rob-v commented:
SortGroupByInfoType didn't cause bug for new mods, it would display unknown types as not sorted, however it isn't general so this enforced ordering - Building, Defense, Infantry - was removed and replaced by sorting by ProductionQueueInfo.Type.
If you know about better general way how to sort it, let me know.

SoScared commented:
Well that was quick.

On stacking, unit production put 'On Hold' takes priority over active production in circumstances which 'on hold' is "closer" to finish compared to active ones. 'On hold' production should thus be prioritized lower than active production.

If possible (polish beyond the issue) the 'unit complete' animation should be removed/skipped from the production info for visual sake, more so with stacked production.

For whatever reason, when re-assigning hotkeys for the info tabs I have to quit and restart the replay in order for the new assignment to take effect.

SoScared commented:
I ran it through some Debug "stress tests" and noticed a few discrepancies https://github.com/OpenRA/OpenRA/issues/5121#issuecomment-289248831

SoScared commented:
Otherwise I gotta say I'm pretty damn stoked to actually get to experience vastly improved visual stats, and for all the mods no less!

SoScared commented:
And just to heap on more work the TD replay clock would likely need adjustment. Example: here's a mockup of the clock 90% of it's original size with the speed indicator positioned as an appendix.
![td3](https://cloud.githubusercontent.com/assets/2715583/24330173/28cf2772-1219-11e7-895b-e94b5fc1d912.png)

rob-v commented:
Thanks a lot for your quick/great test and comments, I was shocked, didn't expect it.
Fixed 1-pixel glitch on the top right corner. Increased spacing between icons and rows. Copied and adjusted .yaml for D2K to have different layout. Added - used ScrollItem/Tab width from .yaml -> removed dependency on last column name - simpler code and customization. 
D2K - adapted (thanks to new .yaml) to more (7?) slots. All mods share same code for Production - so in case of TS change, it should work like for TD (we only need to change e.g. width in .yaml).
Hotkeys - yes, they are assigned at start, otherwise they would need to be retrieved on each key press, so unnecessary usually. Can we keep it?

rob-v commented:
> On production stacking, unit production put 'on hold' takes priority in the slots over active production in circumstances which 'on hold' is closer to finish compared to active ones. Thus 'on hold' production ought to be prioritized lower than active production.

It is currently sorted by remaining time, it means READY, then _in progress or ON HOLD_. You mean you want to display active item before ON HOLD item, even when ON HOLD has less time remaining and when unpaused could finish earlier? So it will be sorted by state (READY, ACTIVE, ON HOLD) and then by time?

SoScared commented:
Let's just say personally I'm highly motivated on the matter plus the predecessor issues dates as far back as early 2014^^

The new spacing between the icons looks great! Now off course that displaced the position vs the support power icons :) I'll also be more specific on the max. likely needed amount of production/support power slots per mod.

RA: 6 production + 5 support powers (6 if Ukraine having all support powers + a capped Chronosphere). I dont believe RA will practically benefit from having 6 slots instead of 5 due to it being infrequent with exception to modded maps. Imo design wise 5 would look the best. Current look:
![ra](https://cloud.githubusercontent.com/assets/2715583/24331220/f4f96770-122f-11e7-82fe-ee560949a110.png)

TD: 9 production + 2 support powers (3 if a faction caps the other factions'  t3 tech) following the logic on RA's slot count. Current look:
![td](https://cloud.githubusercontent.com/assets/2715583/24331326/f19a4d9a-1231-11e7-8ff9-27763545c603.png)

D2K: 7 production + 2 support powers (4 if a faction caps the 2 other factions' t3 tech) following the logic on RA's slot count. Current look:
![d2k](https://cloud.githubusercontent.com/assets/2715583/24331360/e32ef71e-1232-11e7-8dbe-a4bf5196d0e5.png)

TS: 5 production + 3 support powers. Current look :+1: :
![ts](https://cloud.githubusercontent.com/assets/2715583/24331365/06db554a-1233-11e7-8d0b-b6ee6e10f9e5.png)

SoScared commented:
Finally I think the right end side might benefit from a standard space distance due to the show/hide scroll bar (unless there's a custom function to resize the space distance to whether or not there will be a scroll bar). A standard format could look like this:
![end-bar](https://cloud.githubusercontent.com/assets/2715583/24331462/47a32330-1235-11e7-9fd7-390177febc01.png)
With the standard distance between the last slot and the right edge being 32 pixels wide. Without the scroll bar I believe 8 pixels would have been sufficient.

SoScared commented:
> It is currently sorted by remaining time, it means READY, then in progress or ON HOLD. You mean you want to display active item before ON HOLD item, even when ON HOLD has less time remaining and when unpaused could finish earlier? So it will be sorted by state (READY, ACTIVE, ON HOLD) and then by time?

Yeah that sounds about right. Otherwise the production bar would skip a lot if players hold the production on some fascilities while others keep producing. E.g. fast multiple infantry production reveals this well.

SoScared commented:
Finally I think it would be wise for the mod specific production bars to mask out icon slots that surpass the slot limits so as to not garble the display in the fringe cases the production actually surpass the custom slot counts.

SoScared commented:
> So it will be sorted by state (READY, ACTIVE, ON HOLD) and then by time

Actually, if I read this correctly, prioritizing 'READY' behind ACTIVE would eliminate the annoying short flash when the progress bar jumps to the next production in the stacked slots. So maybe (ACTIVE, READY, ON HOLD)?

Thought about removal but it seems like READY is necessary to indicate produced structures that are waiting to be placed.

rob-v commented:
@SoScared I see now, you are not so scared, you scare people. You forced me to calculate all those x+y, pixels here and there in Excel, but it is much better now for me also to adjust all if needed. 
I increased spacing and also adjusted width, but I worked with different x prod. + y support numbers so it was worse at the end.
If there are more prod. icons than available space, they are now skipped. (Lately) In case the mod could provide number of possible/recommended queues, the width could be adjusted automatically, now it has to be done in .yaml manually. I was also thinking about making half size icons to be displayed in the same row when there are more icons needed. Other rows and support powers would stay in current size.
The Paused/Speed indicator was moved to the right of timer.
It is now sorted by ready, active, on hold as you requested, what is really better than always by remaining time.

SoScared commented:
I like to believe I used to scare people. I've been out of consistent practice for over a year so now everyone justl laugh at me:cry:

rob-v commented:
With little training here, it will be fine again.
Fantastic idea - it seems with your idea to sort by active, ready, on hold, the flashing disappeared.

pchote reviewed (changes requested):
The current approach breaks down in multi-queue mods if a player is producing too many unique things at once:

![screen shot 2017-03-26 at 17 18 05](https://cloud.githubusercontent.com/assets/167819/24333000/63f42f9e-1248-11e7-88b2-ace46e0c5dd6.png)

The simplest and IMO clearest way to solve this problem is to give each queue type its own slot (i.e. the TD observer interface will show at most 1 building, one defense, one infantry, etc), and then alternate the icon shown in each slot if more than one thing is being produced (i.e. if a player has two barracks, building 1 minigunner and 1 engineer, then the interface will show the minigunner for 1 second, then the engineer for 1 second, then repeat).  The UI could show a little N / M text indicator below the icon to indicate which queue # the displayed icon is being produced from.

SoScared commented:
Third run-through. I also deliberately increased production on TD to 12 unique units @pchote , the latest commit apparently has that issue sorted out by masking additional production slots over 9:
![td_2](https://cloud.githubusercontent.com/assets/2715583/24333104/e22f32b4-1252-11e7-8be3-bdc8afcc50e1.png)
Support powers would benefit the same feature on all the mods (I can barely see a third support power icon in the above image after granting a faction t3 tech of the opposite faction).

As for the rest:

**RA**:
![ra](https://cloud.githubusercontent.com/assets/2715583/24333128/56e69264-1253-11e7-970f-2f9bb47e47bc.png)
![ra_2](https://cloud.githubusercontent.com/assets/2715583/24333129/597f9afc-1253-11e7-8d4d-4e92de87fb56.png)

**D2K**:
![d2k](https://cloud.githubusercontent.com/assets/2715583/24333134/671f4e6e-1253-11e7-8fd0-1fd8d1b5683f.png)

**TS**:
![ts](https://cloud.githubusercontent.com/assets/2715583/24333136/71290bca-1253-11e7-9d1a-cc1ff6d96e7a.png)

Now as for code implications with future mods I obviously can't speak to that but design wise this is looking pretty AAA to me.

Additional issues regarding UI design might include the game clock position and text style for each mod.

pchote commented:
I really don't like the idea of simply adding more slots to TD, because it just defers the problem (and mods will inevitably find a way to break it) and means that we have a lot of wasted UI space in the common case.

SoScared commented:
Oh no. No no no no I did not wait 3 years for such a A grade stellar opportunity to be dropped with the blink of an eye. Nope! Absolutely negative, not happening.

@rob-v I understand the frustration of having done the significant amount of work in such a tiny time frame and then to have to go through additional levels of reviews. I haven't done any real code-work whatsoever but I think I can relate still with some of my more superfluous PR's in the past. Just stick through the storm until there's a definite conclusion (and then you can rage :smile: ).

SoScared commented:
All right I'm gonna stretch my neurons here to the moon and back so let me see if I can get the argument going forward.

_disclaimer: I know jack about coding._

@pchote So the problem is additional mods coming in with their own production capacity potentially not fitting the related mod's features. Case in point Crystallized Doom's multi queue system providing approximately the same amount of production as the official TD mod, however being tied to the TS mod by default it would inherit the default setup provided for TS thus limiting the usefulness of CD's production info tab's display. Same issue would go for every other mod out there related to any of the official mods.

The alternative is to implement a one-size-fits-all with functionality that would be able to deliver the information uniformly by allowing different units to be displayed with the same slot, solving the need for excessive amount of slots by fitting objects into pre-determined slot types designated as 'infantry', 'vehicles', 'aircraft' etc. Two or more different vehicles produced simultaneously would then be represented by the one slot alternating back and forth between the two until the slot is back onto 1 unit type. The advantage would be for mod creators to have a functional production info bar display available from the get go.

An issue with a uniformed system is the accuracy of portraying real-time information, especially with multi-queues where 3-4 production facilities are represented by a single slot. For units/structures with long build times the result could be satisfactory but the quicker the unit is being produced the harder it would be to follow the information on the info bar. Multiple barracks producing a variety of infantry or production of light vehicles (normally fast built) could be problematic as the slot would likely be noisy at best.

The current proposed setup would demand of mod creators to alter the setup to fit their mods production should their mod deviate on that point vs the official mod it is built upon (or a common neutral version). This would require the modder to learn how to modify the production UI themselves, possibly from a pointer found in some documentation somewhere. This would put a certain amount of stress upon the mod creator depending on the procedure to which the production UI is custom tailored towards the mod and to the extent it would have been modified.

The advantage with having the production UI demand the modder to customize it would be for the UI to present the information tailored to the mods production system and deliver the information plain to the user (assuming that the modder will have to adjust the UI in the first place). Reliable and viewer friendly production UI custom tailored to the specific mod will aid its promotion by familiarizing the user more easily to the units, production style and game progression for the games.

**Uniform production info UI**
:heavy_plus_sign: Easy on modders
:heavy_minus_sign: Limits information

**Custom production info UI**
:heavy_plus_sign: Accurate information
:heavy_minus_sign: Modders needs to engage the UI

SoScared commented:
As I am personally a RA fanatic it wouldn't affect me too much one way or the other (RA using a single queue system) but I'm crossing my fingers be for the production UI to remain customized. I can only hope that there are easy solutions out there regarding UI coding for modders that would make it easier to support that position. IMO the potential far outweighs the costs when it comes to bringing appeal to a standalone mod when you can display its information more accurately when observing a game. Games with obscure, unfamiliar units such as Crystallized Doom would benefit more from having a production UI that at all times presents its production in a plain and straight-forward manner.

SoScared commented:
Regardless, with the current PR its initial solution seem to just have these few polish points left:

- [ ] Mask support-power icons that goes beyond the designated amount of slots (as with production icons).
- [ ] Sort stacked slot priority as (ACTIVE, READY, ON HOLD) to smoothen the stacked production visually.
- [ ] Evaluate text style, clock position in relation to the UI area.

WolfGaming commented:
You can also just add a scrollable thing at the ends if the icons go past the given space can't you? I feel like that would be the best way to tackle the problem with mods being able to break it with a lot of queues.

SoScared commented:
@WolfGaming horizontal scrollbar was mentioned briefly with https://github.com/OpenRA/OpenRA/issues/5121#issuecomment-288826247. I think someone would need to come up with one.

GraionDilach commented:
I suggested the second row for support powers back then because that was feasible, requires no crazy stuff like stacking and can be applied to all mods and even compatible with @pchote's concerns as well.

SoScared commented:
It's not crazy at all, in fact it's being used by one of the most successful modern RTS games the past 8 years:

![dzj84](https://cloud.githubusercontent.com/assets/2715583/24367183/ee0c6baa-131b-11e7-89ac-771456e1732b.jpg)

I wouldn't presume to know the implications (although I try) of such a feature to OpenRA but as an observer tool there can be little doubt of its superiority when it comes to reviewing games and spectating or casting games.
